### PR TITLE
Add PrecisionRecall Op

### LINKED ISSUE #5070: PrecisionRecallEvalutor
- [Evaluator.cpp](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/gserver/evaluators/Evaluator.cpp#L869)
- [precision_recall](http://doc.paddlepaddle.org/develop/doc/api/v2/config/evaluators.html#precision-recall)

### PR DESCRIPTION
**Pull request by pkuyym**
Fixes #5070 

This operator can be used to compute various metrics including:
  - macro average precision
  - macro average recall
  - macro f1 score
  - micro average precision
  - micro average recall
  - micro f1 score

To compute the above metrics, we need to statistic counts for true positives, false positives and false negatives. Here count of true negatives is not necessary, but statisticing it may provide potential usage and the cost is trivial, so the operator also provides count of true negatives.

We define state as a 2-D tensor with shape [class number, 4]. Each row of a state contains statistic variables for corresponding class. Layout of each row is: TP(true positives), FP(false positives), TN(true negatives), FN(false negatives). If 'Input(Weights)' provided, TP, FP, TN, FN will be calculated by given weight instead of instance count.

This operator also supports metrics computing for cross-batch situation. To achieve this, 'Input(StatesInfo)' should be provided. State of current batch data will be accumulated to 'Input(StatesInfo)' and 'Output(AccumStatesInfo)' is the accumulation state.

'Output(BatchMetrics)' is metrics of current batch data while 'Output(AccumStatesInfo)' is metrics of accumulation data.

### CODE REVIEW

File: paddle/operators/precision_recall_op.h [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    const T* predictions_data = in0->data<T>();
+    const int* labels_data = in1->data<int>();
+    const T* weights_data = in2 ? in2->data<T>() : nullptr;
+    const T* states_data = in3 ? in3->data<T>() : nullptr;
+    T* batch_metrics_data = out0->mutable_data<T>(ctx.GetPlace());
+    T* accum_metrics_data = out1->mutable_data<T>(ctx.GetPlace());
+    out2->mutable_data<T>(ctx.GetPlace());
+    auto accum_states = EigenMatrix<T>::From(*out2);
+    accum_states.setZero();
+    T* accum_states_data = out2->data<T>();
```
Comment (typhoonzero) [inline paddle/operators/precision_recall_op.h]: Thought accumulating should be a more general method, like in https://github.com/PaddlePaddle/Paddle/pull/4828#discussion_r147325132 , we can use an accumulating operator to generate more configurable evaluating subnetwork.

File: paddle/operators/precision_recall_op.h [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    auto* out0 = ctx.Output<Tensor>("BatchMetrics");
+    auto* out1 = ctx.Output<Tensor>("AccumMetrics");
+    auto* out2 = ctx.Output<Tensor>("AccumStatesInfo");
+
+    const T* predictions_data = in0->data<T>();
+    const int* labels_data = in1->data<int>();
+    const T* weights_data = in2 ? in2->data<T>() : nullptr;
+    const T* states_data = in3 ? in3->data<T>() : nullptr;
+    T* batch_metrics_data = out0->mutable_data<T>(ctx.GetPlace());
+    T* accum_metrics_data = out1->mutable_data<T>(ctx.GetPlace());
```
Comment (typhoonzero) [inline paddle/operators/precision_recall_op.h]: Outputs of these metrics can just be type `float` or `double`, type `T` should not infect the output type.

File: paddle/operators/precision_recall_op.cc [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+When given 'Input(Predictions)' and 'Input(Labels)', this operator can be used
+to compute various metrics including:
+  - macro average precision
+  - macro average recall
+  - macro f1 score
+  - micro average precision
+  - micro average recall
+  - micro f1 score
+
+To compute the above metrics, we need to statistic counts for true positives,
```
[the remark this answers is not in the document]
Comment (pkuyym) [reply paddle/operators/precision_recall_op.cc]: Done.

File: paddle/operators/precision_recall_op.cc [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+false positives and false negatives. Here count of true negatives is not
+necessary, but statisticing it may provide potential usage and the cost is
```
[the remark this answers is not in the document]
Comment (pkuyym) [reply paddle/operators/precision_recall_op.cc]: Done.

File: paddle/operators/precision_recall_op.h (hunk 2 above)
Comment (pkuyym) [reply paddle/operators/precision_recall_op.h]: Done.

File: paddle/operators/precision_recall_op.h [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+    size_t sample_num = in0->dims()[0];
+    size_t class_dim = in0->dims()[1];
+    size_t state_var_num = 4;  // TP FP TN FN
+
+    // get states info for current batch
+    for (size_t i = 0; i < sample_num; ++i) {
+      size_t max_idx = 0;
+      T max_val = predictions_data[i * class_dim];
+      for (size_t j = 1; j < class_dim; ++j) {
```
[the remark this answers is not in the document]
Comment (typhoonzero) [reply paddle/operators/precision_recall_op.h]: topk will output both probability  and indices.

File: paddle/operators/precision_recall_op.h [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+          typename IndexType = Eigen::DenseIndex>
+using EigenMatrix = framework::EigenMatrix<T, MajorType, IndexType>;
+
+enum StateVariable { TP = 0, FP, TN, FN };
+
+template <typename Place, typename T>
+class PrecisionRecallKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    auto* in0 = ctx.Input<Tensor>("Predictions");
```
Comment (typhoonzero) [inline paddle/operators/precision_recall_op.h]: These temp var names are not quite human-readable.

**Pull request merged**